_posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd

```diff
@@ -0,0 +1,740 @@
+knitr::include_graphics("images/attacker_ds.png")
+```
+
+The other small subset will be used to test the adversary's spying capabilities after training. Let's peek at this one, too:
+
+```{r, eval=TRUE, echo=FALSE, layout="l-body", fig.cap = "Sample from the five-alphabet set used to test the adversary after training(originally: 'background small 2')"}
+knitr::include_graphics("images/actual_test_images.png")
+```
+
+Conveniently, we can use [tfds](https://github.com/rstudio/tfds), the R wrapper to TensorFlow Datasets, to load those subsets:
+
+```{r}
+library(reticulate)
+library(keras)
+library(tfdatasets)
+library(tfautograph)
+library(tfds)
+
+library(purrr)
+
+# we'll use this to train the target model
+# n = 13180
+omni_train <- tfds$load("omniglot", split = "test")
+
+# this is used to train the adversary
+# n = 2720
+omni_spy <- tfds$load("omniglot", split = "small1")
+
+# this we'll use for testing
+# n = 3120
+omni_test <- tfds$load("omniglot", split = "small2")
+
+```
+
+Now first, we train the target model.
+
+### Train target model
+
+The dataset originally has four columns: the image, of size 105 x 105; an alphabet id and a within-dataset character id; and a
+label. For our use case, we're not really interested in the task the target model was/is used for; we just want to get at the
+data. Basically, whatever task we choose, it is not much more than a dummy task. So, let's just say we train the target to
+classify characters *by alphabet*.
+
+We thus throw out all unneeded features, keeping just the alphabet id and the image itself:
+
+```{r}
+# normalize and work with a single channel (images are black-and-white anyway)
+preprocess_image <- function(image) {
+  image %>%
+    tf$cast(dtype = tf$float32) %>%
+    tf$truediv(y = 255) %>%
+    tf$image$rgb_to_grayscale()
+}
+
+# use the first 11000 images for training
+train_ds <- omni_train %>% 
+  dataset_take(11000) %>%
+  dataset_map(function(record) {
+    record$image <- preprocess_image(record$image)
+    list(record$image, record$alphabet)}) %>%
+  dataset_shuffle(1000) %>% 
+  dataset_batch(32)
+
+# use the remaining 2180 records for validation
+val_ds <- omni_train %>% 
+  dataset_skip(11000) %>%
+  dataset_map(function(record) {
+    record$image <- preprocess_image(record$image)
+    list(record$image, record$alphabet)}) %>%
+  dataset_batch(32)
+```
+
+The model consists of two parts. The first is imagined to run in a distributed fashion; for example, on mobile devices (stage
+one). These devices then send model outputs to a central server, where final results are computed (stage two). Sure, you will
+be thinking, this is a convenient setup for our scenario: If we intercept stage one results, we -- most probably -- gain
+access to richer information than what is contained in a model's final output layer. --- That is correct, but the scenario is
+less contrived than one might assume. Just like federated learning [@McMahanMRA16], it fulfills important desiderata: Actual
```

Comment:
I love the word desiderata (I literally have a GitHub repo by that name), but I think it's a bit esoteric in this context. 🤓